proposals/migrate-over-ipfs.md

```diff
@@ -0,0 +1,152 @@
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+10? We're not shipping a new release to brave until we have _a_ way to ship repo migrations to users in China.
+
+## Project definition
+
+#### Brief plan of attack
+
+**Design sketch:**
+
+When migrating a repo:
+
+1. Create a new _temporary_ repo (go-ipfs can't read the current repo because it's an older version).
+    a. If the local go-ipfs node uses a swarm key, skip to step 3.a (download over HTTPs).
+2. Start a new _temporary_ go-ipfs node in the temporary repo.
+    a. This node should not listen for inbound connections as it has no way to know which ports/transports should be configured (can't read the config).
+    b. This node should not expose an API/gateway.
+3. Download the required migration binaries using the temporary go-ipfs node.
+    a. If this fails, download over HTTPs.
+4. Migrate the main go-ipfs node's repo.
+5. Start the main go-ipfs node and _transfer_ (but don't pin) the downloaded migrations into the main repo so that others can download it from this node.
+6. Shut down the temporary node and delete the temporary repo.
+
+Provide the following configuration section:
+
+```js
+{
+    "Migration": {
+        // When true, always use IPFS. When false, never use IPFS. When unset, pick the default
+        // behavior.
+        "UseIPFS": true|false,
+        // When true or unset, use the default gateway. When false, don't use a gateway. When a
+        // string, use the specified gateway.
+        "UseGateway": true|false|"my-gateway",
+        // Whether or not to keep the migration after downloading it.
+        "Keep": "pin"|"cache"|"discard"
+    }
+}
+```
+
+**Implementation steps:**
+
+1. Implement everything but steps 3.a and 5.
+2. Implement step 3.a. We'd need this at a minimum before enabling this feature by default.
+3. Implement step 5. Unless nodes keep around a copy of the migration, this feature isn't going to be all that useful.
+
+#### What does done look like?
+_What specific deliverables should completed to consider this project done?_
+
+1. go-ipfs can download migrations over IPFS (bitswap) without having to contact a centralized server over HTTP.
+2. This feature has been tested by multiple labbers (possibly even tested by a subset of ipfs-desktop users).
+
+####  What does success look like?
+_Success means impact. How will we know we did the right thing?_
+
+Users can start a new version of go-ipfs that switches to a new repo version without downloading anything from a centralized (easy to block) server.
+
+#### Counterpoints &amp; pre-mortem
+_Why might this project be lower impact than expected? How could this project fail to complete, or fail to be successful?_
+
+1. Downloading migrations over IPFS could be too slow/unreliable. Falling back on a gateway should mitigate.
```

Comment:
To make dogfooding safer, we could initially set implicit default to keep HTTPS attempt as the first one and use IPFS as a fallback (for China etc):

```
["dweb.link", "ipfs"]
```
This way the default behavior does not change for most users (migration does not take long time), and we fix edge cases safely.

Reply:
Fair enough.